File: src/main/java/prosia/basarnas/web/util/map/poly/TSNPattern.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package prosia.basarnas.web.util.map.poly;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import org.primefaces.model.map.LatLng;
import org.primefaces.model.map.Marker;
import org.primefaces.model.map.Polyline;
import prosia.basarnas.constant.CommonConstant;
import prosia.basarnas.web.util.map.CommonUtil;

/**
 *
 * @author Ismail
 */
public class TSNPattern implements Serializable {

    private CommonUtil commonUtil = new CommonUtil();
    private LatLng start;
    private double searchLeg;
    private double searchWidth;
    private double searchTrackSpacing;
    private double searchHeading;
    private LatLng pivot;
    private List<LatLng> latlngs;
    private String color;
    private int rotateState;

    @Getter
    private Polyline polyline = new Polyline();
    @Getter
    private List<Marker> markersWaypointToolTip;
    @Getter
    private String contentString;

    public TSNPattern(LatLng start, double searchTrackSpacing, double searchLeg,
            double searchHeading, LatLng pivot, List<LatLng> latlngs, String color) {
        this.color = color;
        this.start = start;
        this.searchTrackSpacing = searchTrackSpacing;
        this.searchLeg = searchLeg;
        this.searchHeading = searchHeading;
        if (pivot == null) {
            this.pivot = new LatLng(start.getLat()
                    + (searchLeg / 2 * CommonConstant.ONE_NM_TO_DEG), start.getLng());
        } else {
            this.pivot = pivot;
        }
        if (latlngs == null || latlngs.isEmpty()) {
            init();
            fristHeadings();
            initWaypointToolTip();
        } else {
            this.latlngs = latlngs;
        }
    }

    private void init() {
        List<LatLng> listLatLngs = new ArrayList<>();
        listLatLngs.add(start);
        LatLng point1 = new LatLng(start.getLat()
                + (this.searchLeg * CommonConstant.ONE_NM_TO_DEG), start.getLng());
        LatLng point2 = new LatLng(point1.getLat(), (point1.getLng()
                + (searchTrackSpacing * CommonConstant.ONE_NM_TO_DEG)));
        LatLng point3 = new LatLng(start.getLat(), point2.getLng());
        LatLng point4 = new LatLng(start.getLat(), (start.getLng()
                - (searchTrackSpacing * CommonConstant.ONE_NM_TO_DEG)));
        LatLng point5 = new LatLng(point2.getLat(), point4.getLng());
        this.rotateState = 0;
        listLatLngs.add(point1);
        listLatLngs.add(point2);
        listLatLngs.add(point3);
        listLatLngs.add(point4);
        listLatLngs.add(point5);
        createAllPolyLine(listLatLngs);
    }

    public void fristHeadings() {
        LatLng firstHeadingPivot = new LatLng(start.getLat(), (start.getLng()
                + (searchTrackSpacing / 2 * CommonConstant.ONE_NM_TO_DEG)));
        List<LatLng> rotation = commonUtil.rotationLatLngs(polyline.getPaths(), firstHeadingPivot, searchHeading);
        polyline.setPaths(rotation);
        pivot = commonUtil.rotationLatLng(firstHeadingPivot, pivot, searchHeading);
    }

    public void headings() {
        List<LatLng> rotation = commonUtil.rotationLatLngs(polyline.getPaths(), pivot, searchHeading);
        polyline.setPaths(rotation);
    }

    public String inFocus() {
        return pivot.getLat() + "," + pivot.getLng();
    }

    public Integer inZoom() {
        return 10;
    }

    public void initWaypointToolTip() {
        markersWaypointToolTip = new ArrayList<>();
        for (int i = 0; i < this.polyline.getPaths().size() - 1; i++) {
            markersWaypointToolTip.add(new Marker(polyline.getPaths().get(i), "Waypoint " + (i + 1)));
        }
    }

    private void createAllPolyLine(List<LatLng> listLatLng) {
        polyline = new Polyline();
        for (LatLng gLatLng : listLatLng) {
            polyline.getPaths().add(new LatLng(gLatLng.getLat(),
                    gLatLng.getLng()));
        }
        polyline.setStrokeColor(color);
        polyline.setStrokeWeight(3);
        polyline.setStrokeOpacity(1);
    }
}

File: src/main/webapp/resources/js/map/poly/TSNPattern.js

/* 
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

function TSNPattern(start, trackSpacing, searchLeg, heading, pivot, latlngs, color){
    this.color = color;
    this.start = start;
    this.trackSpacing = trackSpacing;
    this.searchLeg = searchLeg;
    this.heading = heading;
    if(!pivot){
        this.pivot = new GLatLng(start.lat() + (searchLeg/2 * ONE_NM_TO_DEG), start.lng());
    }else{
        this.pivot = pivot;
    }
    if(latlngs){
        this.l = latlngs;
        this.draw();
    }else{
        this.init();
        this.fristHeadings();
        this.draw();
    }
}

TSNPattern.prototype = new SearchPattern();

TSNPattern.prototype.init = function(){
    this.l = new Array();
    this.l.push(this.start);
    var point1 = new GLatLng(this.start.lat() + (this.searchLeg * ONE_NM_TO_DEG), this.start.lng());
    var point2 = new GLatLng(point1.lat(), (point1.lng() + (this.trackSpacing * ONE_NM_TO_DEG)));
    var point3 = new GLatLng(this.start.lat(), point2.lng());
    var point4 = new GLatLng(this.start.lat(), (this.start.lng() - (this.trackSpacing * ONE_NM_TO_DEG)));
    var point5 = new GLatLng(point2.lat(), point4.lng());
    this.rotateState = 0;
    this.l.push(point1);
    this.l.push(point2);
    this.l.push(point3);
    this.l.push(point4);
    this.l.push(point5);
}

TSNPattern.prototype.fristHeadings = function(){
    rotationGLatLngs(this.l, this.start, this.heading);
    this.pivot = rotationGLatLng(this.start, this.pivot, this.heading);
}

TSNPattern.prototype.headings = function(){
    rotationGLatLngs(this.l, this.pivot, this.heading);
}

TSNPattern.prototype.inFocus = function(){
    map.setCenter(this.pivot, 10);
}

TSNPattern.prototype.move = function(from, to){
    this.moving(from, to);
    this.start = this.l[0];
    this.end = this.l[this.f.length];
    this.pivot = to;
    this.remove();
    this.draw();
}

TSNPattern.prototype.initWaypointToolTip = function(){
    var i, mouseOverCallback, mouseOutCallback;
	if(this.waypointsMarkers){
        for(i in this.waypointsMarkers){
            this.waypointsMarkers[i].setLatLng(this.l[i]);
        }
    }else{
        this.waypointsMarkers = [];
        for(i = 0; i < this.l.length; i++){
			var title = i + 1;
            var waypointMarker = CustomIconMarker(TRANSPARANT_ICON, this.l[i], title.toString());
            mouseOverCallback = function(latlng, marker){
                SEARCH_PATTERN_TOOLTIP.show(latlng, stringHtmlOnMap("Waypoint " + marker.getTitle()));
            };
            mouseOutCallback = function(latlng, marker){
                SEARCH_PATTERN_TOOLTIP.hide();
            };
            measurementPointIntegration(waypointMarker, mouseOverCallback, mouseOutCallback);
            map.addOverlay(waypointMarker);
            this.waypointsMarkers.push(waypointMarker);
        }
    }
}


TSNPattern.prototype.getContentString = function(){
    return "<b class='title'><u>Track Search No-Return Search</u></b>" +
    "<br><b class='field'>Search Leg</b>      : <font class='value'>" + this.searchLeg +  " NM</font>" +
    "<br><b class='field'>Heading </b> : <font class='value'>" + this.heading + " °</font>" +
    "<br><b class='field'>Track Spacing </b> : <font class='value'>" + this.trackSpacing + " NM</font>" +
    "<br><b class='field'>Length</b> : <font class='value'>" + this.getContentStringOfLength() + "</font>";
}